runelite-client/src/main/java/net/runelite/client/ui/ClientFrame.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2018, Woox <https://github.com/wooxsolo>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.ui;
+
+import java.awt.Rectangle;
+import java.awt.event.ComponentAdapter;
+import java.awt.event.ComponentEvent;
+import java.time.Duration;
+import java.time.Instant;
+import javax.swing.JFrame;
+import lombok.Getter;
+
+public class ClientFrame extends JFrame
+{
+	private static final int LARGE_WIDTH_THRESHOLD = 1600;
+	private static final int NEAR_EDGE_DISTANCE = 50;
+	private static final int PANEL_EXPANDED_WIDTH = PluginPanel.PANEL_WIDTH + PluginPanel.SCROLLBAR_WIDTH;
+
+	@Getter
+	private boolean containedInScreen;
+
+	private boolean expanded;
+	private Rectangle oldBounds;
+	private Instant ignoreNextBoundsNullUntil;
+
+	public ClientFrame()
+	{
+		this.addComponentListener(new ComponentAdapter()
+		{
+			@Override
+			public void componentMoved(ComponentEvent e)
+			{
+				super.componentMoved(e);
+
+				if (ignoreNextBoundsNullUntil != null && Instant.now().compareTo(ignoreNextBoundsNullUntil) > 0)
```

Comment:
What's the benefit of unit testing in this case? Seems to me like it would be quicker and easier to test manually.

Reply:
It is always easier to test manually, but unit tests are there so when someone decides to do changes here again it would break tests. So you enforce your logic, plus you are sure that your code works as you intended.

Tho in this case it seems dirty to use time for client resize/move logic

Reply:
Let me know if you can think of a better solution than the "time hack". I've tried some different things like moving `oldBounds = null;` to the setBounds/setLocation/setSize methods, but the problem is that those methods don't run when the user moves/resizes the window and custom chrome is disabled, which is also why I couldn't get the contain feature to work when custom chrome is disabled.

Reply:
I'm still not sure where is the issue that needs timer. Is it the case that some action happens too fast and old values are used?

Reply:
We don't want `oldBounds` to get nulled after running the `expand` method. But if we try to assign to `this.oldBounds` at the end of the method, the component listener events still run after the method finishes and nulls it.